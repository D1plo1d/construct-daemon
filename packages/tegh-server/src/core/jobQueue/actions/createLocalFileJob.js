import path from 'path'
import tmp from 'tmp-promise'
import untildify from 'untildify'

import fs from '../../util/promisifiedFS'
import validateCommandsFileExtension from '../../util/validateCommandsFileExtension'
import Job from '../types/Job'
import JobFile from '../types/JobFile'
import { CREATE_JOB } from './createJob'

const normalize = filePath => path.normalize(untildify(filePath))

const createLocalFileJob = (args) => {
  return async (dispatch, getState) => {

    if (args.localPath === null) {
      throw new Error('localPath must be a string')
    }

    const filePath = normalize(args.localPath)
    const name = path.basename(filePath)

    const stats = await fs.lstatAsync(filePath)

    const localPathConfig = getState().config.printFromLocalPath

    const isAllowedPath = localPathConfig.whitelist.some(validPath => {
      const normalizedValidPath = normalize(validPath)
      return filePath.startsWith(normalizedValidPath)
    })

    /* validation errors */

    if (localPathConfig.enabled !== true) {
      throw new Error(`printing from localPaths is disabled`)
    }

    validateCommandsFileExtension(filePath)

    if (stats.isSymbolicLink() && !localPathConfig.allowSymlinks) {
      throw new Error(`localPath cannot be a symlink`)
    }

    if (!isAllowedPath) {
      return new Error(`localPath is not in a whitelisted directory`)
    }

    /* create the types and actions */
    const job = Job({ name })
    const jobFile = JobFile({
      jobID: job.id,
      name,
      filePath,
      isTmpFile: false,
      quantity: 1,
    })

    const action = {
      type: CREATE_JOB,
      payload: {
        job,
        jobFiles: {
          [jobFile.id]: jobFile,
        },
      },
    }

    return dispatch(action)
  }
}

export default createLocalFileJob
