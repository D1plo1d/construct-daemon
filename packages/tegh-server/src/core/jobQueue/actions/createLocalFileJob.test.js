import path from 'path'
import tmp from 'tmp-promise'

import fs from '../../util/promisifiedFS'
import expectToMatchImmutableSnapshot from '../../util/testing/expectToMatchImmutableSnapshot'
import createLocalFileJob from './createLocalFileJob'

describe('createLocalFileJob', () => {
  it('creates a CREATE_JOB action', async function() {

    const tmpFile = await tmp.file({ postfix: '.gcode' })
    const localPath = tmpFile.path

    const getState = () => ({
      config: {
        printFromLocalPath: {
          enabled: true,
          whitelist: [
            '/otherDirectory',
            path.dirname(localPath),
          ]
        }
      }
    })
    const dispatch = action => action

    let result = await createLocalFileJob({
      localPath
    })(dispatch, getState)

    expectToMatchImmutableSnapshot({
      result,
      redactions: [
        ['payload', 'job', 'name'],
        ['payload', 'job', 'id'],
        ['payload', 'job', 'createdAt'],
        ['payload', 'jobFiles'],
      ],
    })

    const jobFiles = Object.values(result.payload.jobFiles)
    for (const jobFile of jobFiles) {
      expectToMatchImmutableSnapshot({
        result: jobFile,
        redactions: [
          ['name'],
          ['id'],
          ['filePath'],
        ],
      })
      const content = await fs.readFileAsync(jobFile.filePath, 'utf8')
      expect(content).toMatchSnapshot()

      /* clean up tmp file */
      await fs.unlinkAsync(jobFile.filePath)
    }

  })
})
