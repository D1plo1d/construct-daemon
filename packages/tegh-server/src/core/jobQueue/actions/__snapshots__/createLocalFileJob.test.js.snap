// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createLocalFileJob creates a CREATE_JOB action 1`] = `
Object {
  "payload": Object {
    "job": Object {
      "createdAt": "[REDACTED]",
      "id": "[REDACTED]",
      "name": "[REDACTED]",
    },
    "jobFiles": "[REDACTED]",
  },
  "type": "tegh/jobQueue/CREATE_JOB",
}
`;

exports[`createLocalFileJob creates a CREATE_JOB action 2`] = `
Object {
  "filePath": "[REDACTED]",
  "id": "[REDACTED]",
  "isTempFile": false,
  "name": "[REDACTED]",
  "quantity": 1,
}
`;

exports[`createLocalFileJob creates a CREATE_JOB action 3`] = `""`;
